Poor man's simulcast implementation

daniel-abramov opened the pull request:
Per-commit review is suggested (tried to keep the commit history clean and "atomic").

As discussed, I'm submitting this PR to simplify the review of the simulcast that would otherwise have been too overwhelming (over 1000 LOC PR).

This is stage 1 of simulcast implementation (fixes https://github.com/matrix-org/waterfall/issues/31).

daniel-abramov committed:
peer: create configuration for simulcast API

daniel-abramov committed:
peer: accept 3 layers -> subscribe to low always

Implement primitive simulcast without quality switching.

daniel-abramov committed:
simulcast: implement poor man's simulcast

Allow switching the layers by causing a negotiation.

dbkr commented on pkg/conference/data_channel_message_processor.go:
```diff
@@ -11,12 +13,59 @@ func (c *Conference) processTrackSubscriptionDCMessage(
+
+	// Iterate over all published tracks that correspond to the track IDs we want to subscribe to.
+	for id, track := range c.getTracksInfo(toSubscribeTrackIDs) {
+		// Get tre tracks we're subscribed to.
```
```suggestion
		// Get the tracks we're subscribed to.
```

daniel-abramov committed:
simulcast: change the way we treat simulcast

Instead of making it optional (with `nil`), we just create another value
in an enum that would signify the invalid value.

This is not perfect from the category theory point of view, but Go does
not allow ADTs and optional values for that. So we follow the
"convention" that is oftentimes used in such cases.

This would not work if we used more complex data structures, so
generally speaking it's a kludge to simplify handling things in code.

daniel-abramov committed:
simulcast: handle potentially missing layers

daniel-abramov committed:
conference: fix typos

Co-authored-by: David Baker <dbkr@users.noreply.github.com>

SimonBrandner commented on pkg/peer/webrtc_callbacks.go:
```diff
@@ -11,44 +10,33 @@ import (
+	layer, err := RIDToSimulcastLayer(remoteTrack.RID())
 	if err != nil {
-		p.logger.WithError(err).Error("failed to create local track")
+		p.logger.WithError(err).Error("failed to parse RID")
```
We should handle non-simulcast tracks too, right?

daniel-abramov replied:
Yep, I've added this behavior in the upcoming commits (perhaps something went wrong with the rebase), but you probably have already noticed it in the newer commits :)

SimonBrandner commented on pkg/conference/peer_message_processor.go:
```diff
@@ -19,27 +19,46 @@ func (c *Conference) processLeftTheCallMessage(participant *Participant, msg pee
 
+func (c *Conference) processRTPPacketReceivedMessage(participant *Participant, msg peer.RTPPacketReceived) {
+	// For now we just forward the lowest layer always and assume that others don't exist.
+	if msg.Layer != peer.SimulcastLayerLow {
```
Shouldn't it be this instead? Won't the current version forward all but the low layer?

```suggestion
	if msg.Layer == peer.SimulcastLayerLow {
```

daniel-abramov replied:
Yes, the version of that commit indeed only forwarded the low layer and did not allow switching (so in the first iteration I just wrote some sort of a "drop-in" replacement), in the upcoming commit (poor man's implementation) I handled all layers.

SimonBrandner commented on pkg/peer/types.go:
```diff
@@ -47,15 +60,21 @@ func (s SimulcastLayer) String() string {
 	TrackInfo
-	Layer SimulcastLayer
+	// Optional simulcast layer (if any). Would be nil for non-simulcast or audio tracks.
+	Layer *SimulcastLayer
```
I don't like the usage of a pointer here... Could we add `SimulcastLayer.Invalid` or something along those lines?

daniel-abramov replied:
Yes, that's what I did here: https://github.com/matrix-org/waterfall/pull/85/commits/59579be99a6879efbb6a342d2ed864d5147f596b

Although it's a kludge for missing ADTs, it seems like in Go they oftentimes use pointers to signify such things.

SimonBrandner commented on pkg/conference/peer_message_processor.go:
```diff
@@ -19,32 +19,47 @@ func (c *Conference) processLeftTheCallMessage(participant *Participant, msg pee
+	// If it's just a new layer, let's add it to the list of layers of the existing published track.
+	if msg.Layer != nil && !funk.Contains(track.layers, *msg.Layer) {
+		track.layers = append(track.layers, *msg.Layer)
+		participant.publishedTracks[msg.TrackID] = track
```
Won't all the layers have the same `trackId`?

daniel-abramov replied:
They will, that's why in this case I only modify the `layers` without touching the rest of the fields.

SimonBrandner commented on pkg/conference/participant.go:
```diff
@@ -64,3 +64,77 @@ func (p *Participant) sendDataChannelMessage(toSend event.Event) error {
+
+func (p *PublishedTrackInfo) getDesiredLayer(requestedWidth, requestedHeight int) *peer.SimulcastLayer {
+	// Audio track. For them we don't have any simulcast.
+	if p.metadata.isVideoTrack() {
```
```suggestion
	if !p.metadata.isVideoTrack() {
```

daniel-abramov replied:
Yep, was indeed a bug in the initial commit, [fixed it by accident in this commit](https://github.com/matrix-org/waterfall/pull/85/commits/59579be99a6879efbb6a342d2ed864d5147f596b).

SimonBrandner commented on pkg/conference/participant.go:
```diff
@@ -64,3 +64,77 @@ func (p *Participant) sendDataChannelMessage(toSend event.Event) error {
+	desiredLayer := peer.SimulcastLayerLow
+
+	// Determine which simulcast desiredLayer to subscribe to based on the requested resolution.
+	if requestedWidth != 0 || requestedHeight != 0 {
```
Otherwise one could be `0` and therefore `desiredResolution` would too be `0`

```suggestion
	if requestedWidth != 0 && requestedHeight != 0 {
```

daniel-abramov replied:
Thanks! Updated.

daniel-abramov commented on pkg/peer/peer.go:
```diff
@@ -246,3 +217,32 @@ func (p *Peer[ID]) ProcessSDPOffer(sdpOffer string) (*webrtc.SessionDescription,
 
 	return &answer, nil
 }
+
+// Read incoming RTCP packets
+// Before these packets are returned they are processed by interceptors. For things
+// like NACK this needs to be called.
+func (p *Peer[ID]) readRTCP(rtpSender *webrtc.RTPSender) {
+	for {
+		packets, _, err := rtpSender.ReadRTCP()
+		if err != nil {
+			if errors.Is(err, io.ErrClosedPipe) || errors.Is(err, io.EOF) {
+				p.logger.WithError(err).Warn("failed to read RTCP on track")
+				return
+			}
+		}
+
+		// We only want to inform others about PLIs and FIRs. We skip the rest of the packets for now.
+		toForward := []RTCPPacketType{}
+		for _, packet := range packets {
+			// TODO: Should we also handle NACKs?
+			switch packet.(type) {
+			case *rtcp.PictureLossIndication:
+				toForward = append(toForward, PictureLossIndicator)
+			case *rtcp.FullIntraRequest:
+				toForward = append(toForward, FullIntraRequest)
+			}
+		}
+
+		p.sink.Send(RTCPReceived{Packets: toForward, TrackID: rtpSender.Track().ID()})
+	}
+}
```
I think I might have moved it just to keep the order of functions as "first public functions/interface, then private functions", i.e. `WriteRTCP()` is a public function (Go is case sensitive: functions, types, fields that start with capital are exported, whereas those written with a lowercase are internal to the package [folder]).

I should have done it in a separate commit explaining this reasoning to avoid confusion.

daniel-abramov commented on pkg/peer/webrtc_api.go:
```diff
@@ -0,0 +1,54 @@
+
+// Creates Pion's WebRTC API that has all required extensions configured (such as simulcast).
+func createWebRTCAPI() (*webrtc.API, error) {
+	// Enable Extension Headers needed for Simulcast.
```
Oh, you're right, going to remove it!

daniel-abramov commented on pkg/peer/webrtc_api.go:
```diff
@@ -0,0 +1,54 @@
+// Creates a peer connection with a specifically configured API (with simulcast etc).
+func createPeerConnection() (*webrtc.PeerConnection, error) {
+	// TODO: Could we actually reuse the same API configuration for all peer connections?
+	api, err := createWebRTCAPI()
```
Good question: that's why I added the TODO to clarify that :)

The previous implementation from the [draft PR](https://github.com/matrix-org/waterfall/pull/33/files#diff-e83b4d763d27611915a497672c603413fde08249808a76dc627ca661db39c46fR346) created it for each peer connection, so I thought that there must be a reason for that. But if you say that re-using is fine (you probably played more with this API than I did 🙂), then I would change it to re-use the same API.

daniel-abramov committed:
minor: update logging

For brevity, let's not log ICE candidates, it does not carry much of the
information, however let's log all negotiates.

daniel-abramov committed:
channel: increase the size of the channels

This is required since we are going to push lots of messages due to our
peers reporting RTP packets over the channel, so we better not block.

daniel-abramov committed:
simulcast: treat layer requests smarter

So that if a user request layer X and X and lower layers are not
available, we'll send a higher layer. Otherwise the user may get
informed about track published, send a `subscribe` request, but the
information about published layer X has not yet been received by us, so
we would be forced to send nothing to the user when a layer is
requested.

daniel-abramov committed:
peer: handle possible `nil` when reading RTCPs

daniel-abramov committed:
conference: optimize access to the subscriptions

This severely changes the way we treat subscriptions in order to speed
up things. Previous implementation dealt with publish and subscribe
logic in a simple and intuitive way, yet a very inefficient one for
bigger conferences.

Also, it stood on the way of an efficient simulcast implementation.
Since our main goal with simulcast is to make it scale better, this
change is sort of required in a long run although it does not really
bring a lot of performance improvements on top of the small conferences.

Note that in order to keep published and subscribed tracks, we keep them
together with the participants that actually file any changes to avoid
things going awry.

daniel-abramov committed:
participant: move participant logic into package

In order to better separate entities and prevent errors,
participant-related files that now seem to form their own logic, are
moved into their own package to separate entities even more and simplify
making changes.

daniel-abramov committed:
conference: restructure message processing

- Simplify function names a bit (we don't necessary need to mention DC
  in them, it seems like it does not really give much of a context as
  the functino implementation is agnostic from the fact where the
  message was received from: over the data channel, Matrix signaling or
  any other thing).
- Remove separate file that contained processing of data channel only
  messages as all of them are strictly speaking regular peer messages,
  it seems like currently grouping these 100+ LOC together in a separate
  file does not simplify things since it only forces us to jump forwards
  and backwards without separating things (they are still methods on a
  same structure).

daniel-abramov committed:
conference: fix RTCP processing

daniel-abramov committed:
peer: remove unused functions

They became obsolete.

daniel-abramov committed:
conference: fix incorrect adding of tracks

daniel-abramov committed:
simulcast: implement proper RTCP handling

SimonBrandner commented on pkg/peer/types.go:
```diff
@@ -0,0 +1,97 @@
+type ExtendedTrackInfo struct {
+	// Information about a track.
+	TrackInfo
+	// Optional simulcast layer (if any). Would be nil for non-simulcast or audio tracks.
```
Didn't we want to use `SimulcastLayerNone` here?

daniel-abramov replied:
We do. I think I need to fix the outdated comment 🙂

SimonBrandner commented on pkg/peer/types.go:
```diff
@@ -0,0 +1,97 @@
+// Information with some extended (optional fields).
+// Ideally we would want to have different types for different tracks, but Go does not have ADTs,
+// so it's the only way to do it without writing too much of a boilerplate.
+type ExtendedTrackInfo struct {
```
Is there a reason for this to be separate from `TrackInfo`?

daniel-abramov replied:
Hm, good question, I've noticed that `ExtendedTrackInfo` is used in most places. Let me get rid of this abstraction then!

daniel-abramov replied:
Done.

One side effect that I noticed while doing it (and which also reminded me why I introduced it in the first place) is that `PublishedTrack` that contains `TrackInfo` **and** `Layers` (list of published simulcast layers) is slightly confusing in a way that `publishedTrack.Info.Layer` is a useless variable now. But maybe we can ignore it for now as we don't really use it apart single place (which you'll see in a commit that introduced the change). Apart from that it indeed looks simpler.

SimonBrandner commented on pkg/conference/participant/track.go:
```diff
@@ -0,0 +1,100 @@
+	Metadata TrackMetadata
+	// The timestamp at which we are allowed to send the FIR or PLI request. We don't want to send them
+	// too often, so we introduce some trivial rate limiting to not "enforce" too many key frames.
+	canSendKeyframeAt time.Time
```
```suggestion
	canRequestKeyframeAt time.Time
```

daniel-abramov replied:
NB: I just took the name of this field from the previous implementation 🙂 Though I can definitely rename it, requesting probably reflects its meaning better.

daniel-abramov replied:
Done.

SimonBrandner commented on pkg/conference/participant/track.go:
```diff
@@ -0,0 +1,100 @@
+		return desiredLayer
+	}
+
+	// If we wanted high, but high is not available, let's try to see if medium is there.
+	if desiredLayer == peer.SimulcastLayerHigh {
+		if funk.Contains(p.Layers, peer.SimulcastLayerMedium) {
+			return peer.SimulcastLayerMedium
+		}
+
+		// Low is always there, otherwise the `availableLayers` would be empty and we would have returned earlier.
+		return peer.SimulcastLayerLow
+	}
+
+	// If we requested medium and it's not available, we return low (unless the only available layer is high).
+	if desiredLayer == peer.SimulcastLayerMedium {
+		if funk.Contains(p.Layers, peer.SimulcastLayerLow) {
+			return peer.SimulcastLayerLow
+		}
+
+		// Apparently there is only single layer available: high, then we must send it. Maybe others has not yet
+		// been published - the client can always re-request a different quality later if needed.
+		return peer.SimulcastLayerHigh
+	}
+
+	// If we got here, then the low layer was requested, but it's not available.
+	// Let's try to return medium then if it's available.
+	if funk.Contains(p.Layers, peer.SimulcastLayerMedium) {
+		return peer.SimulcastLayerMedium
+	}
+
+	// No other choice rather than sending low.
+	return peer.SimulcastLayerLow
```
I wonder if this logic could be somehow simplified in a for loop, I know I had something along those lines in my original PR, not sure how well it worked/was readable :thinking:

daniel-abramov replied:
Good point: I also don't like this `if` boilerplate here. Previously it was a single `if`, it changed when I realized that the layers can be published in all the combinations: high + low, high + medium, medium + log, high only, etc, handling all of them is not super elegant.

My original logic was that we respect what the user sent to us when requesting the track and use the layer that is desired by the user. However, if a given layer is not available, we must not automatically switch to a different one (higher or lower), but report an error to the user (so that the user can decide if he wants to have a higher layer e.g. if no other layers are available or rather prefer to not have any video at all [for one or another reason]).

daniel-abramov replied:
Ok, rewritten as a loop. The new version is not identical in terms of semantics, but I think it's ok for now given that ideally, we would want to replace this logic with "send an error to the client in case the given quality is not available".

SimonBrandner commented on pkg/conference/participant/tracker.go:
```diff
@@ -0,0 +1,231 @@
+)
+
+type (
+	Subscribers      map[ID]*peer.Subscription
```
```suggestion
	Subscriptions      map[ID]*peer.Subscription
```

SimonBrandner commented on pkg/conference/participant/tracker.go:
```diff
@@ -0,0 +1,231 @@
+
+// Adds a new track to the list of published tracks, i.e. by calling it we inform the tracker that there is new track
+// that has been published and that we must take into account from now on.
+func (t *Tracker) AddTrack(
```
```suggestion
func (t *Tracker) PublishTrack(
```

daniel-abramov replied:
But we don't publish the track here: the way has been posted at this point already, we add it to the tracker, so we start tracking its state and keep it in sync with the participants' state. Another reason why I chose `AddTrack()` here is to be consistent and symmetric with other functions: we have `add participant`, `remove participant`, `for each participant` and we also have `add track`, `remove track` and `for each track`.

Otherwise having only `add track` called `publish track` might be confusing (i.e. why `publish track`, but `remove track` etc).

SimonBrandner replied:
Right, what about `AddPublishedTrack()`? I wanted there to be a clear distinction between subscribing and adding a published track

daniel-abramov replied:
Ah, but there is no function to mix it with subscribing, i.e. for subscribing there is a `Subscribe` function in a tracker, so adding tracks (`AddTrack()`) is only for published tracks (which we also explain in a comment).

I renamed it to `PublishedTrack` for each function, i.e. `AddTrack` is now `AddPublishedTrack`, and `UpdateTrackMetadata` is now `UpdatePublishedTrackMetadata`. Let me know if it's better (to me it seems a bit too lengthy now, but I don't have a strong opinion on that 🙂).

SimonBrandner replied:
> I renamed it to `PublishedTrack` for each function, i.e. `AddTrack` is now `AddPublishedTrack`, and `UpdateTrackMetadata` is now `UpdatePublishedTrackMetadata`. Let me know if it's better (to me it seems a bit too lengthy now, but I don't have a strong opinion on that ).

Thank you! It's way more readable for me now

SimonBrandner reviewed (commented):
I have to admit that some of the abstractions are becoming a bit unclear to me again (e.g. `Subscription` vs `SubscribedTrack`). I wonder if the naming could be improved here

daniel-abramov commented:
> I have to admit that some of the abstractions are becoming a bit unclear to me again (e.g. Subscription vs SubscribedTrack). I wonder if the naming could be improved here

We don't have a `SubscribedTrack` structure if I'm not mistaken 🤔, but we do have a `Subscription` which refers to a single subscription (subscription to a single track).

Or did you mean that you would prefer `Subscription` called `SubscribedTrack`? - I thought that it might be confusing because it would mean that `SubscribedTrack` encapsulates a track to which someone is subscribed (to me it feels like it gives a notion that there is one `SubscribedTrack`, i.e. if `asdf` is a track, to which someone is subscribed, does it have to be a type of `SubscribedTrack` and stored in a map, so that we can differentiate tracks to which someone is subscribed vs tracks that no-one is subscribed to).

Whereas with `Subscription` it's just a **subscription** (has a connotation that it is a **new** entity) to an _existing_ published track, which means that there could be many of them, i.e. we may have 3,4 or 10 subscriptions to a single existing track.

SimonBrandner commented:
> > I have to admit that some of the abstractions are becoming a bit unclear to me again (e.g. Subscription vs SubscribedTrack). I wonder if the naming could be improved here
> 
> We don't have a `SubscribedTrack` structure if I'm not mistaken thinking, but we do have a `Subscription` which refers to a single subscription (subscription to a single track).
> 
> Or did you mean that you would prefer `Subscription` called `SubscribedTrack`? - I thought that it might be confusing because it would mean that `SubscribedTrack` encapsulates a track to which someone is subscribed (to me it feels like it gives a notion that there is one `SubscribedTrack`, i.e. if `asdf` is a track, to which someone is subscribed, does it have to be a type of `SubscribedTrack` and stored in a map, so that we can differentiate tracks to which someone is subscribed vs tracks that no-one is subscribed to).
> 
> Whereas with `Subscription` it's just a **subscription** (has a connotation that it is a **new** entity) to an _existing_ published track, which means that there could be many of them, i.e. we may have 3,4 or 10 subscriptions to a single existing track.

Ah, yes, the whole `Subscription`, `Subscribers` and `TrackSubscribers` has me confused...

daniel-abramov committed:
tracker: rename subscription types for clarity

daniel-abramov committed:
peer: get rid of `ExtendedTrackInfo`

Replace it with `TrackInfo` with an additional `layer` field.

daniel-abramov committed:
refactor: rename RTCP key frame interval var

daniel-abramov committed:
refactor: rename `*Track*` to `*PublishedTrack*`

daniel-abramov committed:
peer: rework (simplify) RTCP processing

Workaround the errors caused by the RTCP handling.

daniel-abramov committed:
refactor: move `types` into `common` package

daniel-abramov committed:
peer: create mutex protected peer state

daniel-abramov committed:
peer: use state to search for tracks

It seems like otherwise it does not quite work as expected due to
`GetReceivers()` not returning simulcast tracks! Perhaps a bug in Pion.

SimonBrandner commented on pkg/common/track_info.go:
```diff
@@ -0,0 +1,95 @@
+	}
+}
+
+type ConnectionWrapper struct {
+	connection *webrtc.PeerConnection
+}
+
+func NewConnectionWrapper(connection *webrtc.PeerConnection) ConnectionWrapper {
+	return ConnectionWrapper{
+		connection: connection,
+	}
+}
+
+func (c ConnectionWrapper) Subscribe(track *webrtc.TrackLocalStaticRTP) (*webrtc.RTPSender, error) {
+	return c.connection.AddTrack(track)
+}
+
+func (c ConnectionWrapper) Unsubscribe(sender *webrtc.RTPSender) error {
+	return c.connection.RemoveTrack(sender)
+}
```
This looks like it should be in a different file? Also, should the `c`s be pointers?

daniel-abramov replied:
👌 Moved.

SimonBrandner commented on pkg/peer/messages.go:
```diff
@@ -40,19 +46,6 @@ type DataChannelMessage struct {
-	Type    RTCPPacketType
-	Content rtcp.Packet
+type KeyFrameRequestReceived struct {
+	common.TrackInfo
```
Is this a PLI, should it be named that way? What about FIRs?

daniel-abramov replied:
So, since RTCPs [failed a lot](https://github.com/matrix-org/waterfall/issues/87) and resulted in errors, I decided that why don't we just simplify it then by converting all read RTCPs into just some "keyframe request" (so both PLIs and FIRs are just sent as keyframe request messages from the `Peer` so that it's up to the peer to then convert the "keyframe request" in what it thinks makes sense. Currently, all "keyframe requests" are indeed converted into PLIs upon arrival to the concrete peer.

This solved all the RTCP errors that were sent (so it seems like these were indeed FIRs that we tried to forward [might require more work to be properly handled]). Though the freezes in Firefox (and only in Firefox) still happen from time to time, but there are no grey frames anymore 🙂

SimonBrandner commented on pkg/peer/state/peer_state.go:
```diff
@@ -0,0 +1,72 @@
+	return p.remoteTracks[RemoteTrackId{id, simulcast}]
+}
+
+func (p *PeerState) GetRemoteTracks() []*webrtc.TrackRemote {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	tracks := make([]*webrtc.TrackRemote, 0, len(p.remoteTracks))
+	for _, track := range p.remoteTracks {
+		tracks = append(tracks, track)
+	}
+
+	return tracks
+}
```
Seems to be unused

daniel-abramov replied:
Yep, I just created it for completeness (since it's in a separate package), but I can remove unused functions. The idea of the `PeerState` was to isolate mutex-protected data in its own structure to prevent users from accessing the fields without locking the mutex. Essentially, we isolate it to ensure that we access things in a thread-safe way.

SimonBrandner replied:
Yeah, I think we should avoid unused methods...

daniel-abramov replied:
👌 Removed.

SimonBrandner reviewed (commented):
I think we're almost there!

daniel-abramov committed:
simulcast: simpler desired layer fallback

The complexity in that place is not worth it actually.

daniel-abramov committed:
peer: move connection wrapper in its own file

daniel-abramov committed:
peer(state): remove unused function

SimonBrandner reviewed (approved):
LGTM

Do we wait to resolve the Firefox issue before merging?

@dbkr, do you want to have another look?

daniel-abramov committed:
refactor: move subscription into its own package

daniel-abramov committed:
subscription: create a watchdog for RTP frames

daniel-abramov commented:
> LGTM
> 
> Do we wait to resolve the Firefox issue before merging?
> 
> @dbkr, do you want to have another look?

Yep, the issue with grey frames seems to be a bug, I need to fix it either prior to merging or after that.
